src/OgAccess.php

```diff
@@ -0,0 +1,258 @@
+
+      //static::setPermissionCache($group_entity, $user, TRUE, $permissions);
+    }
+
+    if (!$skip_alter && !isset($post_alter_cache[$operation])) {
+      // Let modules alter the permissions. So we get the original ones, and
+      // pass them along to the implementing modules.
+      // @todo: Check if still needed to do a clone, since the cache is static
+      // we don't want it to be altered.
+      $alterable_permissions = static::getPermissionsCache($group_entity, $user, TRUE);
+      $context = array(
+        'operation' => $operation,
+        'group' => $group_entity,
+        'user' => $user,
+      );
+      $cacheable_metadata = new CacheableMetadata;
+      \Drupal::moduleHandler()->alter('og_user_access', $alterable_permissions, $cacheable_metadata, $context);
+
+      static::setPermissionCache($group_entity, $user, FALSE, $alterable_permissions, $cacheable_metadata);
+    }
+
+    $altered_permissions = static::getPermissionsCache($group_entity, $user, FALSE);
+
+    $user_is_group_admin = !empty($altered_permissions['permissions'][static::ADMINISTER_GROUP_PERMISSION]);
+    if (($user_is_group_admin && !$ignore_admin) || !empty($altered_permissions['permissions'][$operation])) {
+      // User is a group admin, and we do not ignore this special permission
+      // that grants access to all the group permissions.
+      return AccessResult::allowed()->addCacheableDependency($altered_permissions['cacheable_metadata']);
+    }
+    else {
+      return AccessResult::forbidden();
+    }
+  }
+
+  /**
+   * Check if a user has access to a permission on a certain entity context.
+   *
+   * @param string $operation
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity object.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   (optional) The user object. If empty the current user will be used.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   Returns TRUE if the user has access to the permission, otherwise FALSE, or
+   *   if the entity is not in OG context, function _will return NULL. This allows
+   *   a distinction between FALSE - no access, and NULL - no access as no OG
+   *   context found.
+   */
+  public static function userAccessEntity($operation, EntityInterface $entity, AccountInterface $user = NULL) {
+    $result = AccessResult::neutral();
+
+    // Entity isn't saved yet.
+    if ($entity->isNew()) {
+      return $result->addCacheableDependency($entity);
+    }
+
+    $entity_type = $entity->getEntityTypeId();
+    $bundle = $entity->bundle();
+
+    $is_group_content = Og::isGroupContent($entity_type, $bundle);
+
+    if (Og::isGroup($entity_type, $bundle)) {
+      $user_access = static::userAccess($entity, $operation, $user);
+      if ($user_access->isAllowed()) {
+        return $user_access;
+      }
+      else {
+        // An entity can't be a group and group content in the same time. The
```

Comment:
@chx I'll change it back to the original comment, as an entity _can_ be a group and a group content -- this is what allows having sub-groups.